graylog2-server/src/main/java/org/graylog2/rest/resources/entities/preferences/EntityListPreferencesResource.java

```diff
@@ -0,0 +1,94 @@
+import org.graylog2.plugin.database.ValidationException;
+import org.graylog2.rest.resources.entities.preferences.model.EntityListPreferences;
+import org.graylog2.rest.resources.entities.preferences.model.StoredEntityListPreferences;
+import org.graylog2.rest.resources.entities.preferences.service.EntityListPreferencesService;
+
+import javax.inject.Inject;
+import javax.validation.constraints.NotEmpty;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+import static org.graylog2.shared.rest.documentation.generator.Generator.CLOUD_VISIBLE;
+
+@RequiresAuthentication
+@Api(value = "EntityLists", description = "Entity Lists Preferences", tags = {CLOUD_VISIBLE})
+@Path("/entitylists/preferences")
+public class EntityListPreferencesResource {
+
+    private final EntityListPreferencesService entityListPreferencesService;
+
+    @Inject
+    public EntityListPreferencesResource(final EntityListPreferencesService entityListPreferencesService) {
+        this.entityListPreferencesService = entityListPreferencesService;
+    }
+
+    @POST
+    @Path("/{entity_list_id}")
+    @Timed
+    @ApiOperation(value = "Create or update user preferences for certain entity list")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @NoAuditEvent("Storage details have not yet been decided") //TODO
+    public Response create(@ApiParam(name = "JSON body", required = true) EntityListPreferences entityListPreferences,
+                           @ApiParam(name = "entity_list_id", required = true) @PathParam("entity_list_id") @NotEmpty String entityListId,
+                           @Context UserContext userContext) throws ValidationException {
+
+        final String currentUserId = userContext.getUserId();
+        final StoredEntityListPreferences storedPreferences = new StoredEntityListPreferences(currentUserId, entityListId, entityListPreferences);
+        entityListPreferencesService.save(storedPreferences);
+        return Response.ok().build();
+    }
+
+    @GET
+    @Path("/{entity_list_id}")
+    @Timed
+    @ApiOperation(value = "Get preferences for user's entity list", response = EntityListPreferences.class)
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiResponses(value = {
+            @ApiResponse(code = 404, message = "Preferences not found.")
+    })
+    public EntityListPreferences get(@ApiParam(name = "entity_list_id", required = true) @PathParam("entity_list_id") @NotEmpty String entityListId,
+                                     @Context UserContext userContext) throws NotFoundException {
+
+        final String currentUserId = userContext.getUserId();
+        final StoredEntityListPreferences entityListPreferences = entityListPreferencesService.get(currentUserId, entityListId);
```

Comment:
Could we give these some types, so they are not both string-string? It's a source of errors, easy to flip the order and hard to discover. Thanks!!

Reply:
What about a builder instead of artificial types?

Reply:
I have solved it with a separate class for complex id, with builder.